teps/0052-tekton-results-automated-run-resource-cleanup.md

```diff
@@ -46,17 +46,48 @@ keep track of.
 
 Today, most users view Tekton Run history via the Tekton Pipeline API (e.g.
 kubectl, tkn, dashboard, etc). However since all CRD objects are
-[stored in etcd](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/),
-this creates a problem for the Pipeline controller - as the build history grows
-over time, so does the corpus of data the controller needs to handle in its data
-plane. Over time this grows unbounded, significantly degrading the performance
-of the cluster as more Runs are completed.
+[stored in the Kubernetes API](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/),
+the data plane that is used for managing Runs by the controller is also the
+source of data users use for historical data.
+
+Over time this grows unbounded as more Runs are completed. For Tekton, this is
+particularly aggravated due to the automated nature of Runs being created.
+
+This has increased Pipeline controller latency due to `List` calls made:
+
+- [TaskRuns](https://github.com/tektoncd/pipeline/blob/75f3776da39f9440aada89c3e176e7db16af3121/pkg/reconciler/pipelinerun/pipelinerun.go#L991)
+- [PipelineRuns](https://github.com/tektoncd/pipeline/blob/75f3776da39f9440aada89c3e176e7db16af3121/pkg/reconciler/pipelinerun/pipelinerun.go#L998)
+- [Pods](https://github.com/tektoncd/pipeline/blob/75f3776da39f9440aada89c3e176e7db16af3121/pkg/reconciler/taskrun/taskrun.go#L392)
```

Comment:
Yeah I can file an issue. 

Funny enough, embedding status info is why these queries are being done in the first place - it's a mechanism to prevent double scheduling if a Run resource is interrupted without the status info being written back (i.e. query for things with the parent label, if it already exists use that, else create what is missing).
We might be able to avoid some of the query lookups with some journaling, but this will likely expose us to different complexity (I'm primarily worried about name collisions). 😬

Reply:
https://github.com/tektoncd/pipeline/issues/3812

Reply:
> it's a mechanism to prevent double scheduling if a Run resource is interrupted without the status info being written back (i.e. query for things with the parent label, if it already exists use that, else create what is missing)

ah okay, so this is because for all of these cases there are (at least) 2 things that have to be done:

1. create the resource (taskrun/pod/run)
2. update the status of the pipelinerun/task run with that resource

and there's no guarantee (2) will always happen if (1) happened.

i see what you mean - probably not something we can fix then X'D

sounds pretty solid! but last question: can you confirm that list calls to the API server grow in time when more objects exist in the cluster, even when they are being filtered by label? (this is something that could probably be proven without needing to use any controllers, etc., you could make a script that creates + lists objects - or just point to a blog where someone already does this XD)

Reply:
Here's some initial data!

Took your advice and used a new CRD. Code can be found here: https://github.com/wlynch/experimental/blob/crd-loadtest/crd-loadtest/main.go

### User latency

This is pretty noticeable. With ~7k objects we were reaching ~2s list latencies using a label selection query that should return 1 object:

![image](https://user-images.githubusercontent.com/1844673/110865282-63226780-8291-11eb-9e94-454ba57ade5c.png)

### Controller latency

It's also worth looking at what lister performance looks like, to simulate the object caching that's done on controllers. 
Results look promising, but keep in mind these results are showing cached responses - a cache miss would likely be subject to the same latencies as we see above. This probably means we don't need to worry controller performance for now, even if the non-cached flow most humans experience is degraded.

![image](https://user-images.githubusercontent.com/1844673/110869208-fc547c80-8297-11eb-81cb-0aab3b3aa9ab.png)

Reply:
Nice, thanks for doing this investigation!

I'm not seeing an update in the TEP itself, am I missing something? i.e. the TEP still talks about controller list latencies, and has a section on "Be smarter about how Tekton Controllers query resources"

Reply:
(ps thanks for your patience @wlynch  as when i demanded that you file an issue for something that in the end made no sense and then gradually came around to seeing this clearly 🙏 🤣 )

Reply:
Updated! PTAL.